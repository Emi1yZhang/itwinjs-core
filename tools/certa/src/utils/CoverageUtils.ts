/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
import * as fs from "fs";
import * as path from "path";
import * as uuid from "uuid";
import { onExit, spawnChildProcess } from "./SpawnUtils";

/**
 * Helper function for relaunching (as a child process) the current process running under `nyc` for measuring code coverage.
 * Once that completes, a second `nyc` child process will be spawned to report the coverage results.
 * Returns a promise that will be resolved with either 0 or the first non-zero child process exit code, once the reporter process terminates.
 */
export async function relaunchForCoverage(): Promise<number> {
  const nyc = require.resolve("c8/bin/c8");
  const relaunchArgs = [
    nyc,
    "--config",
    path.join(process.cwd(), ".nycrc"),
    "--",
    ...process.argv,
  ];

  // By splitting "instrument/runTests" and "report coverage" into two steps, we allow test runners the option of
  // running separate (concurrent) instrumented processes that also write to `nyc`'s temp directory.
  const instrumentedProcess = spawnChildProcess("node", relaunchArgs, { ...process.env, NYC_CWD: process.cwd() }, false, path.join(process.cwd(), "../.."));
  return await onExit(instrumentedProcess);
}

/**
 * Writes nyc or istanbul-generated coverage data to a uniquely-named file inside `nyc`'s temp directory.
 * All such files in the nyc temp directory are merged together when nyc reports are generated via `nyc report`.
 * @param coverageData The value of an nyc/istanbul-generated `__coverage__` object.
 */
export function writeCoverageData(coverageData: any): void {
  const nycTempDirAbsolute = process.env.NODE_V8_COVERAGE!;
  if (!fs.existsSync(nycTempDirAbsolute))
    throw new Error(`Cannot save coverage data - nyc temp directory "${nycTempDirAbsolute}" does not exist.`);

  // Use uuid/v4 to generate a unique filename, just like `nyc` does.
  const coverageFileName = path.join(nycTempDirAbsolute, `${uuid.v4()}.json`);
  fs.writeFileSync(coverageFileName, JSON.stringify(coverageData));
}
