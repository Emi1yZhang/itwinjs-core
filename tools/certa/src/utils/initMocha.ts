/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
// NB: This file is not a CommonJs module - it needs to run in the browser. Do not import or export modules here!

type CertaConfig = import("../CertaConfig").CertaConfig;
declare const _CERTA_CONFIG: CertaConfig; // eslint-disable-line @typescript-eslint/naming-convention
declare const _CERTA_MOCHA_HOOKS: any; // eslint-disable-line @typescript-eslint/naming-convention

((config: CertaConfig) => {
  const mochaOpts = config.mochaOptions;

  // This is essentially equivalent to `mocha.setup("bdd")`, except it works in both node and browser environments.
  mocha.ui("bdd");
  mocha.suite.emit("pre-require", typeof (window) === "undefined" ? global : window, null, mocha);

  mocha.reporter(mochaOpts.reporter, mochaOpts.reporterOptions);
  // TODO: Come back and fix useColors to color
  (mocha as any).color(true);
  mocha.timeout(mochaOpts.timeout);
  if (mochaOpts.fgrep)
    mocha.fgrep(mochaOpts.fgrep);
  if (mochaOpts.grep)
    mocha.grep(mochaOpts.grep);
  if (mochaOpts.invert)
    mocha.invert();
  if (mochaOpts.forbidOnly)
    mocha.forbidOnly();

  // Disable timeouts when debugging.
  if (config.debug)
    mocha.timeout(0);

  if (_CERTA_MOCHA_HOOKS) {
    const { mochaOptions, mochaGlobalSetup, mochaGlobalTeardown } = _CERTA_MOCHA_HOOKS;
    if (mochaOptions)
      mochaOptions();
    if (mochaGlobalSetup)
      mocha.globalSetup(mochaGlobalSetup);
    if (mochaGlobalTeardown)
      mocha.globalTeardown(mochaGlobalTeardown);
  }

})(_CERTA_CONFIG);
