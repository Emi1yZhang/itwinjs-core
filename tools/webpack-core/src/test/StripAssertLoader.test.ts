/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/* eslint-disable @typescript-eslint/no-var-requires, @typescript-eslint/naming-convention */
import { Volume } from "memfs";
import { expect } from "chai";
import * as path from "path";
import * as fs from "fs-extra";
import { clearCache, clearFileSystem, fsFromJson, getTestConfig, runWebpack } from "./TestUtils";
import { setApplicationDir } from "../utils/paths";

describe("strip-assert-loader", () => {
  let testConfig: any;
  const vol = new Volume();

  let loaderContents: string;
  const loaderPath = path.join(__dirname, "../loaders/strip-assert-loader.js");

  const loaderRule = {
    test: () => true,
    loader: loaderPath,
    enforce: "pre",
  };

  before(() => {
    setApplicationDir(path.join(__dirname, "assets/strip-assert-loader"));
    loaderContents = fs.readFileSync(loaderPath, "utf8");
  });

  it("should strip cjs asserts from core-bentley package", async () => {
    vol.fromJSON({
      [loaderPath]: loaderContents,
      "lib/test/assets/strip-assert-loader-test/test.js": `const core_bentley_1 = require("@itwin/core-bentley");\n(0, core_bentley_1.assert)(false);`,
    });
    testConfig = getTestConfig("assets/strip-assert-loader-test/test.js", [], ["@itwin/core-bentley"], [loaderRule]);

    const result = await runWebpack(testConfig, vol);
    expect(result.errors.length).to.equal(0, result.errors.join("\n"));
    expect(result.warnings.length).to.equal(0, result.warnings.join("\n"));

    const bundleContents = vol.readFileSync(path.join(result.outputPath, "test.js"), "utf8");
    expect(bundleContents).to.matchSnapshot();
    expect(bundleContents).to.not.contain(`core_bentley_1.assert`); // Sanity check - make sure nobody blindly updates snapshot :)
  });

  it("should strip esm asserts from core-bentley package", async () => {
    vol.fromJSON({
      [loaderPath]: loaderContents,
      "lib/test/assets/strip-assert-loader-test/test.js": `import { assert } from "@itwin/core-bentley";\nassert("THIS WAS A TEST ASSERTION");`,
    });
    testConfig = getTestConfig("assets/strip-assert-loader-test/test.js", [], ["@itwin/core-bentley"], [loaderRule]);

    const result = await runWebpack(testConfig, vol);
    expect(result.errors.length).to.equal(0, result.errors.join("\n"));
    expect(result.warnings.length).to.equal(0, result.warnings.join("\n"));

    const bundleContents = vol.readFileSync(path.join(result.outputPath, "test.js"), "utf8");
    expect(bundleContents).to.matchSnapshot();
    expect(bundleContents).to.not.contain("THIS WAS A TEST ASSERTION"); // Sanity check - make sure nobody blindly updates snapshot :)
  });

  it("should safely strip cjs asserts inside single-line if", async () => {
    vol.fromJSON({
      [loaderPath]: loaderContents,
      "lib/test/assets/strip-assert-loader-test/test.js": `const core_bentley_1 = require("@itwin/core-bentley");\nif (true) (0, core_bentley_1.assert)(false);`,
    });
    testConfig = getTestConfig("assets/strip-assert-loader-test/test.js", [], ["@itwin/core-bentley"], [loaderRule]);

    const result = await runWebpack(testConfig, vol);
    expect(result.errors.length).to.equal(0, result.errors.join("\n"));
    expect(result.warnings.length).to.equal(0, result.warnings.join("\n"));

    const bundleContents = vol.readFileSync(path.join(result.outputPath, "test.js"), "utf8");
    expect(bundleContents).to.matchSnapshot();
    expect(bundleContents).to.not.contain(`core_bentley_1.assert`); // Sanity check - make sure nobody blindly updates snapshot :)
  });

  it("should safely strip esm asserts inside single-line if", async () => {
    vol.fromJSON({
      [loaderPath]: loaderContents,
      "lib/test/assets/strip-assert-loader-test/test.js": `import { assert } from "@itwin/core-bentley";if (true)\nassert("THIS WAS A TEST ASSERTION");`,
    });
    testConfig = getTestConfig("assets/strip-assert-loader-test/test.js", [], ["@itwin/core-bentley"], [loaderRule]);

    const result = await runWebpack(testConfig, vol);
    expect(result.errors.length).to.equal(0, result.errors.join("\n"));
    expect(result.warnings.length).to.equal(0, result.warnings.join("\n"));

    const bundleContents = vol.readFileSync(path.join(result.outputPath, "test.js"), "utf8");
    expect(bundleContents).to.matchSnapshot();
    expect(bundleContents).to.not.contain("THIS WAS A TEST ASSERTION"); // Sanity check - make sure nobody blindly updates snapshot :)
  });

  it("should strip cjs asserts from _within_ core-bentley package", async () => {
    vol.fromJSON({
      [loaderPath]: loaderContents,
      "lib/test/assets/strip-assert-loader-test/test.js": `require("@itwin/core-bentley");`,
      "lib/test/assets/strip-assert-loader-test/node_modules/@itwin/core-bentley/package.json": `{"main": "lib/cjs/index.js"}`,
      "lib/test/assets/strip-assert-loader-test/node_modules/@itwin/core-bentley/lib/cjs/index.js": `const Assert_1 = require("./Assert");\n(0, Assert_1.assert)("THIS WAS A TEST ASSERTION");`,
      "lib/test/assets/strip-assert-loader-test/node_modules/@itwin/core-bentley/lib/cjs/Assert.js": `function assert(foo, bar) {console.log(foo,bar);} exports.assert = assert;`,
    });
    testConfig = getTestConfig("assets/strip-assert-loader-test/test.js", [], [], [loaderRule]);

    const result = await runWebpack(testConfig, vol);
    expect(result.errors.length).to.equal(0, result.errors.join("\n"));
    expect(result.warnings.length).to.equal(0, result.warnings.join("\n"));

    const bundleContents = vol.readFileSync(path.join(result.outputPath, "test.js"), "utf8");
    expect(bundleContents).to.matchSnapshot();
    expect(bundleContents).to.not.contain("THIS WAS A TEST ASSERTION"); // Sanity check - make sure nobody blindly updates snapshot :)
  });

  it("should strip esm asserts from _within_ core-bentley package", async () => {
    vol.fromJSON({
      [loaderPath]: loaderContents,
      "lib/test/assets/strip-assert-loader-test/test.js": `import "@itwin/core-bentley";`,
      "lib/test/assets/strip-assert-loader-test/node_modules/@itwin/core-bentley/package.json": `{"module": "lib/esm/index.js"}`,
      "lib/test/assets/strip-assert-loader-test/node_modules/@itwin/core-bentley/lib/esm/index.js": `import { assert } from "./Assert";\nassert("THIS WAS A TEST ASSERTION");`,
      "lib/test/assets/strip-assert-loader-test/node_modules/@itwin/core-bentley/lib/esm/Assert.js": `export function assert(foo, bar) {console.log(foo,bar);}`,
    });
    testConfig = getTestConfig("assets/strip-assert-loader-test/test.js", [], [], [loaderRule]);

    const result = await runWebpack(testConfig, vol);
    expect(result.errors.length).to.equal(0, result.errors.join("\n"));
    expect(result.warnings.length).to.equal(0, result.warnings.join("\n"));

    const bundleContents = vol.readFileSync(path.join(result.outputPath, "test.js"), "utf8");
    expect(bundleContents).to.matchSnapshot();
    expect(bundleContents).to.not.contain("THIS WAS A TEST ASSERTION"); // Sanity check - make sure nobody blindly updates snapshot :)
  });

  it("should safely handle commented-out asserts", async () => {
    vol.fromJSON({
      [loaderPath]: loaderContents,
      "lib/test/assets/strip-assert-loader-test/test.js": `import { assert } from "@itwin/core-bentley";\n// assert("THIS WAS A TEST ASSERTION");\n/* assert(false); */`,
    });
    testConfig = getTestConfig("assets/strip-assert-loader-test/test.js", [], ["@itwin/core-bentley"], [loaderRule]);

    const result = await runWebpack(testConfig, vol);
    expect(result.errors.length).to.equal(0, result.errors.join("\n"));
    expect(result.warnings.length).to.equal(0, result.warnings.join("\n"));

    const bundleContents = vol.readFileSync(path.join(result.outputPath, "test.js"), "utf8");
    expect(bundleContents).to.matchSnapshot();
  });

  afterEach(() => {
    vol.reset();
    clearFileSystem(__dirname);
    clearCache(__dirname);
  });
});
