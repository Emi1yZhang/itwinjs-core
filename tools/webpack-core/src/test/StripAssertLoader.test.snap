// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`strip-assert-loader should safely handle commented-out asserts 1`] = `
"(window[\\"webpackJsonp\\"] = window[\\"webpackJsonp\\"] || []).push([[0],[
/* 0 */
/***/ (function(module, exports, __webpack_require__) {

module.exports = __webpack_require__(1);


/***/ }),
/* 1 */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

\\"use strict\\";
__webpack_require__.r(__webpack_exports__);
/* harmony import */ var _itwin_core_bentley__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(2);
/* harmony import */ var _itwin_core_bentley__WEBPACK_IMPORTED_MODULE_0___default = /*#__PURE__*/__webpack_require__.n(_itwin_core_bentley__WEBPACK_IMPORTED_MODULE_0__);

// void 0;                             
/* void 0;        */

/***/ }),
/* 2 */
/***/ (function(module, exports) {

module.exports = require(\\"@itwin/core-bentley\\");

/***/ })
],[[0,1]]]);"
`;

exports[`strip-assert-loader should safely strip cjs asserts inside single-line if 1`] = `
"(window[\\"webpackJsonp\\"] = window[\\"webpackJsonp\\"] || []).push([[0],[
/* 0 */
/***/ (function(module, exports, __webpack_require__) {

module.exports = __webpack_require__(1);


/***/ }),
/* 1 */
/***/ (function(module, exports, __webpack_require__) {

const core_bentley_1 = __webpack_require__(2);
if (true) void 0;                           

/***/ }),
/* 2 */
/***/ (function(module, exports) {

module.exports = require(\\"@itwin/core-bentley\\");

/***/ })
],[[0,1]]]);"
`;

exports[`strip-assert-loader should safely strip esm asserts inside single-line if 1`] = `
"(window[\\"webpackJsonp\\"] = window[\\"webpackJsonp\\"] || []).push([[0],[
/* 0 */
/***/ (function(module, exports, __webpack_require__) {

module.exports = __webpack_require__(1);


/***/ }),
/* 1 */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

\\"use strict\\";
__webpack_require__.r(__webpack_exports__);
/* harmony import */ var _itwin_core_bentley__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(2);
/* harmony import */ var _itwin_core_bentley__WEBPACK_IMPORTED_MODULE_0___default = /*#__PURE__*/__webpack_require__.n(_itwin_core_bentley__WEBPACK_IMPORTED_MODULE_0__);
if (true)
void 0;                             

/***/ }),
/* 2 */
/***/ (function(module, exports) {

module.exports = require(\\"@itwin/core-bentley\\");

/***/ })
],[[0,1]]]);"
`;

exports[`strip-assert-loader should strip cjs asserts from _within_ core-bentley package 1`] = `
"(window[\\"webpackJsonp\\"] = window[\\"webpackJsonp\\"] || []).push([[0],[
/* 0 */
/***/ (function(module, exports, __webpack_require__) {

module.exports = __webpack_require__(1);


/***/ }),
/* 1 */
/***/ (function(module, exports, __webpack_require__) {

__webpack_require__(2);

/***/ }),
/* 2 */
/***/ (function(module, exports, __webpack_require__) {

const Assert_1 = __webpack_require__(3);
void 0;                                           

/***/ }),
/* 3 */
/***/ (function(module, exports) {

function assert(foo, bar) {console.log(foo,bar);} exports.assert = assert;

/***/ })
],[[0,1]]]);"
`;

exports[`strip-assert-loader should strip cjs asserts from core-bentley package 1`] = `
"(window[\\"webpackJsonp\\"] = window[\\"webpackJsonp\\"] || []).push([[0],[
/* 0 */
/***/ (function(module, exports, __webpack_require__) {

module.exports = __webpack_require__(1);


/***/ }),
/* 1 */
/***/ (function(module, exports, __webpack_require__) {

const core_bentley_1 = __webpack_require__(2);
void 0;                           

/***/ }),
/* 2 */
/***/ (function(module, exports) {

module.exports = require(\\"@itwin/core-bentley\\");

/***/ })
],[[0,1]]]);"
`;

exports[`strip-assert-loader should strip esm asserts from _within_ core-bentley package 1`] = `
"(window[\\"webpackJsonp\\"] = window[\\"webpackJsonp\\"] || []).push([[0],[
/* 0 */
/***/ (function(module, exports, __webpack_require__) {

module.exports = __webpack_require__(1);


/***/ }),
/* 1 */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

\\"use strict\\";
// ESM COMPAT FLAG
__webpack_require__.r(__webpack_exports__);

// CONCATENATED MODULE: ./lib/test/assets/strip-assert-loader-test/node_modules/@itwin/core-bentley/lib/esm/Assert.js
function assert(foo, bar) {console.log(foo,bar);}
// CONCATENATED MODULE: ./lib/test/assets/strip-assert-loader-test/node_modules/@itwin/core-bentley/lib/esm/index.js

void 0;                             
// CONCATENATED MODULE: ./lib/test/assets/strip-assert-loader-test/test.js


/***/ })
],[[0,1]]]);"
`;

exports[`strip-assert-loader should strip esm asserts from core-bentley package 1`] = `
"(window[\\"webpackJsonp\\"] = window[\\"webpackJsonp\\"] || []).push([[0],[
/* 0 */
/***/ (function(module, exports, __webpack_require__) {

module.exports = __webpack_require__(1);


/***/ }),
/* 1 */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

\\"use strict\\";
__webpack_require__.r(__webpack_exports__);
/* harmony import */ var _itwin_core_bentley__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(2);
/* harmony import */ var _itwin_core_bentley__WEBPACK_IMPORTED_MODULE_0___default = /*#__PURE__*/__webpack_require__.n(_itwin_core_bentley__WEBPACK_IMPORTED_MODULE_0__);

void 0;                             

/***/ }),
/* 2 */
/***/ (function(module, exports) {

module.exports = require(\\"@itwin/core-bentley\\");

/***/ })
],[[0,1]]]);"
`;
